### ISSUE TITLE
Global type parameter and reflection

### ISSUE THREAD

**Issue by csyonghe** [shader-slang/slang#944]
This is another issue with our current API, so it may longer be relevant if we are going to surface the new API. 

The scenario is if I have a shader module with global generic types, and I want to "compile" the module to get the reflection data, I will get a compilation error complaining that the types for generic parameters are not provided, even when no entry points are requested in the CompilationRequest.

We may or may not fix this for the old API, but I still want to take a note of it here.

**Comment by tfoleyNV** [comment]
Yeah, that sounds like a bug.

I haven’t gone and removed global generic type parameters yet simply because none of the alternatives are fully usable for all the same scenarios.

**Comment by csyonghe** [comment]
I was not saying that we should remove `type_param` just yet. In fact, I find it very convenient when writing shaders that are generic on input vertex attributes. An alternative to the test case in #947 is to define `TVertexFormat` as an entry point generic argument, but that requires the same generic argument be defined on both the vertex and fragment entry point, and the application must ensure that the compatible specializations are used together. Or we will require something that bundles together a bunch of entry points in a generic struct type, which we tried to avoid all the times for its own complexities.

**Comment by tfoleyNV** [comment]
> Or we will require something that bundles together a bunch of entry points in a generic struct type, which we tried to avoid all the times for its own complexities.

That's exactly what I'd like to see us support sooner or later: allow entry points to be declared in `struct` types, and then allow for those `struct`s to be generic to express type parameters that need to be shared across entry points.

With the revamped approach to handling entry points that we have in the compiler internals today, I *think* that allowing for entry points as member functions should not be too much added complexity now (but it still requires work since there has been zero implementation effort on that).

The gotcha that I think we've run into before when we've discussed replacing global-scope parameters with `struct` members, or generic parameters on `struct`s is that we ultimately have better tools for expressing composition at the global scope than we do for `struct` types. When you `import` a module you instantly get access to any parameters and code declared in that module, and its (transitive) dependencies automatically get loaded as well, with no need to worry about double-importing when there are "diamond" patterns in the module dependency graph.

In contrast, without a good implementation of inheritance, all we can do with `struct` types is composition via aggregation, which either leads to duplication when there are diamond patterns in the dependency graph, or else requires us to pass around a lot more state explicitly when calling into aggregated sub-objects.

Long story short: in my mind moving to shader entry points nested in `struct` types is an important step, and the next step after that is to implement some kind of inheritance/trait/mixin system for composing such `struct` types, at which point there shouldn't be much left that requires the global scope.

**Comment by csyonghe** [comment]
Closing the issue as global type parameters are no longer relavent.

As suggested here, we need a better component system for top-level assembly.
